tests/unit/s2n_session_cache_callback_blocked.c

```diff
@@ -0,0 +1,437 @@
+        result = 11;
+    }
+
+    shutdown_rc = -1;
+    while(shutdown_rc != 0) {
+        shutdown_rc = s2n_shutdown(conn, &blocked);
+    }
+
+    s2n_connection_free(conn);
+
+    /* Give the server a chance to avoid sigpipe */
+    sleep(1);
+
+    /* Session resumption with bad session state */
+    conn = s2n_connection_new(S2N_CLIENT);
+    s2n_connection_set_read_fd(conn, readfd);
+    s2n_connection_set_write_fd(conn, writefd);
+
+    /* Change the format of the session state and check we cannot deserialize it */
+    serialized_session_state[0] = 3;
+    if (s2n_connection_set_session(conn, serialized_session_state, serialized_session_state_length) == 0) {
+        result = 12;
+    }
+
+    if (s2n_errno != S2N_ERR_INVALID_SERIALIZED_SESSION_STATE) {
+        result = 13;
+    }
+
+    serialized_session_state[0] = 0;
+    /* Change the protocol version (36th byte) in session state */
+    if (serialized_session_state_length < 36) {
+        result = 14;
+    }
+
+    serialized_session_state[35] = 30;
+    if (s2n_connection_set_session(conn, serialized_session_state, serialized_session_state_length) < 0) {
+        result = 15;
+    }
+
+    if (s2n_negotiate(conn, &blocked) == 0) {
+        result = 16;
+    }
+
+    if (s2n_errno != S2N_ERR_BAD_MESSAGE) {
+        result = 17;
+    }
+
+    s2n_connection_free(conn);
+    s2n_config_free(config);
+
+    /* Give the server a chance to avoid sigpipe */
+    sleep(1);
+
+    _exit(result);
+}
+
+int main(int argc, char **argv)
+{
+    struct s2n_connection *conn;
+    struct s2n_config *config;
+    s2n_blocked_status blocked;
+    int status;
+    pid_t pid;
+    int server_to_client[2];
+    int client_to_server[2];
+    char *cert_chain_pem;
+    char *private_key_pem;
+    char buffer[256];
+    int bytes_read;
+    int shutdown_rc = -1;
+
+    /* init session cache lock field, which is used to mock a remote
+     * connection/event block
+     */
+    for (int i = 0; i < 256; i++) {
+        session_cache[i].lock = 1;
+    }
+
+    BEGIN_TEST();
+    EXPECT_NOT_NULL(cert_chain_pem = malloc(S2N_MAX_TEST_PEM_SIZE));
+    EXPECT_NOT_NULL(private_key_pem = malloc(S2N_MAX_TEST_PEM_SIZE));
+
+    EXPECT_SUCCESS(setenv("S2N_ENABLE_CLIENT_MODE", "1", 0));
+
+    /* Create a pipe */
+    EXPECT_SUCCESS(pipe(server_to_client));
+    EXPECT_SUCCESS(pipe(client_to_server));
+
+    /* Create a child process */
+    pid = fork();
+    if (pid == 0) {
+        /* This is the child process, close the read end of the pipe */
+        EXPECT_SUCCESS(close(client_to_server[0]));
+        EXPECT_SUCCESS(close(server_to_client[1]));
+
+        /* Write the fragmented hello message */
+        mock_client(client_to_server[1], server_to_client[0]);
+    }
+
+    /* This is the parent */
+    EXPECT_SUCCESS(close(client_to_server[1]));
+    EXPECT_SUCCESS(close(server_to_client[0]));
+
+    /* initial handshake */
+    {
+        int *count = malloc(sizeof(int));
+        *count = 1;
```

Comment:
why do you need `count`? Doesn't seem to be used anywhere

Reply:
good question! I don't see it being used anywhere. removed it.